src/sage/graphs/graph_decompositions/slice_decomposition.pyx

```diff
@@ -0,0 +1,970 @@
+        for k in range(nneighbors):
+            u_int = neighbors[k]
+            j = deref(sigma_inv)[u_int] # get the position of u
+            if j <= i:
+                continue # already taken care of
+
+            if lex_label != NULL:
+                deref(lex_label)[j].push_back (v_int)
+
+            p = part_of[j] # get the part of u
+            l = part_head[p] # get the beginning of the part containing u
+
+            # if not last and next elem belongs in the same part (ie #part >= 2)
+            if l < n - 1 and part_of[l + 1] == p:
+                if l != j: # not already first elem of the part
+                    # Place u at the position of the head of the part
+                    t_int = sigma[l]
+                    deref(sigma_inv)[t_int], deref(sigma_inv)[u_int] = j, l
+                    sigma[j], sigma[l] = t_int, u_int
+                    if lex_label != NULL:
+                        swap[vector[int]](deref(lex_label)[j],
+                                          deref(lex_label)[l])
+                    j = l
+                part_head[p] += 1 # move the head of the part to next elem
+
+            # if part p was not already cut in two during this iteration, we
+            # create a new part using subpart
+            if subpart[p] < old_nparts:
+                subpart[p] = nparts
+                part_head[nparts] = j
+                if xslice_len != NULL:
+                    part_len[nparts] = 0
+                subpart[nparts] = 0
+                nparts += 1
+
+            # Finally, we update the name of the part for position j and set v
+            # as predecessor of u
+            part_of[j] = subpart[p]
+            if xslice_len != NULL:
+                part_len[p] -= 1
+                part_len[subpart[p]] += 1
+            if pred != NULL:
+                deref(pred)[j] = i
+
+    if need_to_delete_sigma_inv:
+        del sigma_inv
+
+def slice_decomposition(G, initial_vertex=None):
+    r"""
+    Compute a slice decomposition of the simple undirected graph
+
+    INPUT:
+
+    - ``G`` -- a Sage graph.
+
+    - ``initial_vertex`` -- (default: ``None``); the first vertex to consider.
+
+    OUTPUT:
+
+    An object of type :class:`~sage.graphs.graph_decompositions.slice_decomposition.SliceDecomposition`
+    that represents a slice decomposition of ``G``
+
+    .. NOTE::
+
+        Loops and multiple edges are ignored during the computation of the slice
+        decomposition.
+
+    ALGORITHM:
+
+    The method use the algorithm based on "partition refinement" described in
+    [HMPV2000]_ and [TCHP2008]_.
+    The time complexity of this algorithm is in `O(n + m)`, and our
+    implementation follows that complexity for ``SparseGraph``. For
+    ``DenseGraph``, the complexity is `O(n^2)`.
+
+    EXAMPLES:
+
+    Slice decomposition of the Petersen Graph::
+
+        sage: G = graphs.PetersenGraph()
+        sage: SD = G.slice_decomposition(); SD
+        [0[1[4[5]]][2[6]][3][9][7][8]]
```

Comment:
You may add spaces between items to improve the readability.

Reply:
done in 1fbb00ebdb1de31ab09729fe20f6e949acefd770